### ISSUE TITLE
error when attempting concurrent writes

### ISSUE THREAD

**Issue by nathandao** [neo4jrb/neo4j#724]
When trying to write to the database with 2 or more workers from a resque job, all the jobs fail with this error:
LockClient[....] can't wait on resource RWLock[....]

However, the job succeeds if I only use 1 worker

Here is the resque job:

<pre><code>class WordProcessJob
  @queue = :word_process_job

  class &lt;&lt;self

    def perform(data)
      words = data['words']
      url = data['url']
      process(words, url)
    end

    private

      def process(words, url)
        word_data = construct_word_data(words)
        cypher = construct_cypher(word_data, url)
        Neo4j::Session.query(cypher)
      end

      def construct_word_data(words)
        word_data = '['
        first = true
        words.each do | word, frequency |
          word = word.to_s.downcase
          if first == true
            first = false
          else
            word_data += ','
          end
          word_data += "{word:'#{word}', frequency: #{frequency}}"
        end
        word_data += ']'
        return word_data
      end

      def construct_cypher(word_data, url)
        cypher = "match (s:Site { url: '#{url}' })
          foreach(n IN #{word_data}|
            merge (w:Word { word: n.word })
            merge (w)-[r:appeared_in]->(s)
            on create set r.frequency = n.frequency
            on match set r.frequency = r.frequency + n.frequency
          )"
        return cypher
      end
    #end private
  end
end</code></pre>

**Comment by subvertallchris** [comment]
Handling of this and other temporary, retryable error was added to Neo4j-core in PR 171, https://github.com/neo4jrb/neo4j-core/pull/171, but hasn't been part of a release yet. You should use the master branch of that and this gem, bound to a recent stable commit since both gems are preparing for a new release and breaking changes might be added at some point. In my projects, I'm using the following commits and haven't run intro problems:

```
gem 'neo4j', github: 'neo4jrb/neo4j', ref: '34a93a9a2cd40e319b7531eff8cf68e8a069f84d'
gem 'neo4j-core', github: 'neo4jrb/neo4j-core', ref: '5ed2ea999a10ab21c1faf65d0b86bd41dfdbf4ff'
```

The feature added by the PR linked above lets you set the number of retries using an environment var, `ENV['NEO4J_RETRY_COUNT']`. It defaults to 10 but I have it set to 25 in my apps and it's working nicely.

You can minimize your chances of encountering this by improving Cypher query performance. Read up on using params, you will see a huge improvement if you do. You may also want to use Neo4j 2.2.0 RC01 if possible, though I think the official release is imminent.

**Comment by subvertallchris** [comment]
I ran into an issue that may be related. Multiple sessions starting together were trying to create the same constraints that already existed, resulting in locks on the DB. This fixes that. Make sure you're using `5c70da6a5eddaab51ff23b8b492657aabb662da4` from core.

**Comment by subvertallchris** [comment]
Merged with ref `7c1afdedffd15dc1ed353a639ba2246b021ca6e7`

**Comment by nathandao** [comment]
Hi,
Thanks for the help! HUGE performance boost after applying all suggestions and switching to neo4j enterprise 2.2.
Cheers!
